pkg/slo-controller/noderesource/plugins/rdmadevicereource/plugin.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rdmadeviceresource
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/koordinator-sh/koordinator/apis/configuration"
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/slo-controller/noderesource/framework"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+const PluginName = "RDMADeviceResource"
+
+const (
+	ResetResourcesMsg  = "reset node rdma resources"
+	UpdateResourcesMsg = "node rdma resources from device"
+
+	NeedSyncForResourceDiffMsg = "rdma resource diff is big than threshold"
+)
+
+var (
+	ResourceNames = []corev1.ResourceName{
+		extension.ResourceRDMA,
+	}
+)
+
+var client ctrlclient.Client
+
+type Plugin struct{}
+
+func (p *Plugin) Name() string {
+	return PluginName
+}
+
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;patch
+// +kubebuilder:rbac:groups=core,resources=nodes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups=scheduling.koordinator.sh,resources=devices,verbs=get;list;watch
+// +kubebuilder:rbac:groups=topology.node.k8s.io,resources=noderesourcetopologies,verbs=get;list;watch;create;update
+
+func (p *Plugin) Setup(opt *framework.Option) error {
+	client = opt.Client
+
+	opt.Builder = opt.Builder.Watches(&schedulingv1alpha1.Device{}, &DeviceHandler{})
+
+	return nil
+}
+
+func (p *Plugin) NeedSync(strategy *configuration.ColocationStrategy, oldNode, newNode *corev1.Node) (bool, string) {
+	for _, resourceName := range ResourceNames {
+		klog.V(4).Info("rdma IsResourceDiff start")
```

Comment:
1. need to use a higher verbosity for debugging messages, e.g. V(6)
2. need to log the node name and the resource name